src/hashset.h

```diff
@@ -0,0 +1,205 @@
+    void (*rehashingCompleted)(hashset *t);
+    /* Allow a hashset to carry extra caller-defined metadata. The extra memory
+     * is initialized to 0. */
+    size_t (*getMetadataSize)(void);
+    /* Flag to disable incremental rehashing */
+    unsigned instant_rehashing : 1;
+    /* Allow the caller to store some data here in the type. It's useful for the
+     * rehashingStarted and rehashingCompleted callbacks. */
+    void *userdata;
+} hashsetType;
+
+typedef enum {
+    HASHSET_RESIZE_ALLOW = 0,
+    HASHSET_RESIZE_AVOID,
+    HASHSET_RESIZE_FORBID,
+} hashsetResizePolicy;
+
+typedef void (*hashsetScanFunction)(void *privdata, void *element);
+
+/* Constants */
+#define HASHSET_BUCKET_SIZE 64 /* bytes */
+
+/* Scan flags */
+#define HASHSET_SCAN_EMIT_REF (1 << 0)
+#define HASHSET_SCAN_SINGLE_STEP (1 << 2)
+
+typedef struct {
+    hashset *hashset;
+    long index;
+    int table;
+    int posInBucket;
+    /* unsafe iterator fingerprint for misuse detection. */
+    uint64_t fingerprint;
+    int safe;
+} hashsetIterator;
+
+typedef struct hashsetStats {
+    int htidx;
+    unsigned long buckets;       /* num buckets */
+    unsigned long maxChainLen;   /* probing chain length */
+    unsigned long totalChainLen; /* buckets with probing flag */
+    unsigned long htSize;        /* buckets * positions-per-bucket */
+    unsigned long htUsed;        /* num elements */
+    unsigned long *clvector;
+} hashsetStats;
+
+/* --- Prototypes --- */
+
+/* Hash function (global seed) */
+void hashsetSetHashFunctionSeed(const uint8_t *seed);
+uint8_t *hashsetGetHashFunctionSeed(void);
+uint64_t hashsetGenHashFunction(const char *buf, size_t len);
+uint64_t hashsetGenCaseHashFunction(const char *buf, size_t len);
+
+/* Global resize policy */
+void hashsetSetResizePolicy(hashsetResizePolicy policy);
+
+/* Hashset instance */
+hashset *hashsetCreate(hashsetType *type);
+void hashsetRelease(hashset *t);
+void hashsetEmpty(hashset *t, void(callback)(hashset *));
+hashsetType *hashsetGetType(hashset *t);
+void *hashsetMetadata(hashset *t);
+size_t hashsetSize(hashset *t);
+size_t hashsetBuckets(hashset *t);
+size_t hashsetProbeCounter(hashset *t, int table);
+size_t hashsetMemUsage(hashset *t);
+void hashsetPauseAutoShrink(hashset *t);
+void hashsetResumeAutoShrink(hashset *t);
+int hashsetIsRehashing(hashset *t);
+int hashsetIsRehashingPaused(hashset *t);
+void hashsetRehashingInfo(hashset *t, size_t *from_size, size_t *to_size);
+int hashsetRehashMicroseconds(hashset *s, uint64_t us);
+int hashsetExpand(hashset *t, size_t size);
+int hashsetTryExpand(hashset *t, size_t size);
+int hashsetExpandIfNeeded(hashset *t);
+int hashsetShrinkIfNeeded(hashset *t);
+hashset *hashsetDefragInternals(hashset *t, void *(*defragfn)(void *));
+
+/* Elements */
+int hashsetFind(hashset *t, const void *key, void **found);
+void **hashsetFindRef(hashset *t, const void *key);
+/* void *hashsetFetchElement(hashset *t, const void *key); */
```

Comment:
I added it when working over kvstore and other stuff, but it seemed not to be needed in the end. I'll delete it.

(It's like hashsetFind but with a different interface and it requires the element to be present. Otherwise we can't store zeroes/nulls in the table, which is a boring limitation.)